### PR TITLE
measure_repeats_and_dynamics

### PR DESCRIPTION
**Pull request by RomanPudashkin**
Resolves: https://github.com/musescore/MuseScore/issues/15232
Resolves: https://github.com/musescore/MuseScore/issues/17680

### CODE REVIEW

**Comment by cbjeukendrup** [comment]
In this case: 
Scherm­afbeelding 2023-12-20 om 18 01 26
The measure with measure repeat starts `ppp`, but shouldn't it start `fff`? 
I think this is a debatable case... not sure what the expected behaviour should be.

File: src/engraving/playback/playbackcontext.cpp [hunk 1]
```diff
@@ -293,6 +303,51 @@ void PlaybackContext::handleAnnotations(const ID partId, const Segment* segment,
...
+    }
+
+    auto endIt = mu::findLess(m_dynamicsMap, rangeEndTick);
+    if (endIt != m_dynamicsMap.end()) {
+        endIt = std::next(endIt);
+    }
+
+    DynamicMap newDynamics;
+    for (auto it = startIt; it != endIt; ++it) {
+        int tick = copyStartTick + it->first;
```
Comment (cbjeukendrup) [inline src/engraving/playback/playbackcontext.cpp:344]:
```suggestion
        int tick = copyStartTick + (it->first - rangeStartTick);
```
to make it work correctly if the original measure is not at the start of the score

Comment (RomanPudashkin) [reply src/engraving/playback/playbackcontext.cpp]: Thanks! Fixed

File: src/engraving/playback/playbackcontext.cpp [hunk 2]
```diff
@@ -293,6 +303,61 @@ void PlaybackContext::handleAnnotations(const ID partId, const Segment* segment,
...
+        }
+    }
+}
+
+void PlaybackContext::copyDynamicsInRange(const int rangeStartTick, const int rangeEndTick, const int newDynamicsOffsetTick)
+{
+    auto startIt = m_dynamicsMap.find(rangeStartTick);
+    if (startIt == m_dynamicsMap.end()) {
+        return;
+    }
```
Comment (cbjeukendrup) [inline src/engraving/playback/playbackcontext.cpp:345]: This combination of `find` and `return` seems to break this case: 
Scherm­afbeelding 2023-12-20 om 21 01 14
(when it's not the first measure)

It looks like replacing `find` with `lower_bound` gives the correct result.

File: src/engraving/playback/playbackcontext.cpp [hunk 3]
```diff
@@ -293,6 +303,61 @@ void PlaybackContext::handleAnnotations(const ID partId, const Segment* segment,
...
+
+    auto endIt = mu::findLess(m_dynamicsMap, rangeEndTick);
+    if (endIt != m_dynamicsMap.end()) {
+        endIt = std::next(endIt);
+    }
```
Comment (cbjeukendrup) [inline src/engraving/playback/playbackcontext.cpp:350]: Is this equivalent to `m_dynamicsMap.lowerBound(rangeEndTick)`?

**Comment by cbjeukendrup** [comment]
Shouldn't we do the same for playing techniques? (Possibly as a separate issue/PR)

**Review by cbjeukendrup — APPROVED** [review]

**Comment by DmitryArefiev** [comment]
Tested on Win10. There are some issues:

1. 4 measure repeats don't work correctly

https://github.com/musescore/MuseScore/assets/90187801/a5aafd5f-ec01-4992-9c3b-bb1d2a52732d

2. Strange thing, 1 measure repeats stop working after adding repeat barlines before


https://github.com/musescore/MuseScore/assets/90187801/353915c1-3157-4ba0-8f06-116d6b5ab5ef

**Comment by DmitryArefiev** [comment]
Tested on Win10. 

> 
Strange thing, 1 measure repeats stop working after adding repeat barlines before

This **still occurs**


https://github.com/musescore/MuseScore/assets/90187801/bf57a09b-ac22-4714-9d16-25b6658e2f11


Attaching the scores [Scores.zip](https://github.com/musescore/MuseScore/files/15078141/Scores.zip)

**Comment by DmitryArefiev** [comment]
Tested #15232 #17680 on Win10, Mac13.6 - **FIXED**    

This case is https://github.com/musescore/MuseScore/pull/20598#issuecomment-2072433674 also fixed

**Pull request merged**